apps/src/lab2/redux/predictLevelRedux.ts

```diff
@@ -0,0 +1,62 @@
+import {sendPredictLevelReport} from '@cdo/apps/code-studio/progressRedux';
+import {PayloadAction, createSelector, createSlice} from '@reduxjs/toolkit';
+import {setLoadedPredictResponse} from '@cdo/apps/lab2/lab2Redux';
+import {RootState} from '@cdo/apps/types/redux';
+
+export interface PredictLevelState {
+  response: string;
+  hasSubmittedResponse: boolean;
+}
+
+const initialState: PredictLevelState = {
+  // User's response for the level, if the level is a predict level. It is an empty string if this
+  // is not a predict level or if the user has not yet submitted a response.
+  response: '',
+  // If the user has submitted a predict response for the current level.
+  hasSubmittedResponse: false,
+};
+
+// SELECTORS
+
+// The predict answer is locked if the level does not allow multiple predict attempts
+// and the user has already submitted a response.
+export const isPredictAnswerLocked = createSelector(
+  [
+    (state: RootState) =>
+      state.lab.levelProperties?.predictSettings?.allowMultipleAttempts,
+    (state: RootState) => state.predictLevel.hasSubmittedResponse,
+  ],
+  (allowMultipleAttempts, hasSubmittedResponse) => {
+    return !allowMultipleAttempts && hasSubmittedResponse;
+  }
+);
+
+// REDUCER
+const predictSlice = createSlice({
+  name: 'predictLevel',
+  initialState,
+  reducers: {
+    setPredictResponse(state, action: PayloadAction<string>) {
+      state.response = action.payload;
+    },
+    setHasSubmittedPredictResponse(state, action: PayloadAction<boolean>) {
```

Comment:
Non-blocking nit: Would it make sense to simplify these names, like `setResponse` and `setSubmitted`? (I'm not attached to this, and I'm not even sure it's better. I wonder if `setHasSubmittedPredictResponse` is getting into "hard to parse" territory.)

Reply:
I want the names to be understandable when importing into another file, so I think I was to keep `setPredictResponse` as-is. Looking at the code again, I actually am not using `setHasSubmittedPredictResponse` anywhere because that value is always set via on of the extra reducers. So I'll just delete that one :)